### PR TITLE
types: allow calling `SchemaType.cast()` without `parent` and `init` parameters

### LINKED ISSUE #14748: Optional Arguments in SchemaTypes.path().cast required by Typescript
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

7.6.7

### Node.js version

18.20.3

### MongoDB server version

7.0.12

### Typescript version (if applicable)

5.2.2

### Description

It looks like the typescript signature for SchemaTypes.path().cast require some (what appear to be) optional fields.

I wanted to create a subdocument instance without creating a model. With reference to https://github.com/Automattic/mongoose/issues/9076,

I had 
image

which works exactly as intended with the //@ts-ignore, but otherwise gives me a type error. Upon further examination,
it appears that 
image
doc and init are required fields in the typescript signature, whereas they seem to be optional in a lot of the documentation I've seen.

This is my top-level schema file:
image

Please let me know if I'm mistaken somewhere, many thanks!! 

### Steps to Reproduce

Please see above. 

### Expected Behavior

Optional (?) fields should be required by typescript signature

### PR DESCRIPTION
**Pull request by vkarpov15**
Fix #14748
Re: #9076

**Summary**

The `doc` and `init` parameters to `cast()` should be optional in TypeScript, we have plenty of tests that call `SchemaType.prototype.cast()` with only 1 arg. Plus #9076 recommends using `Schema.Types.Subdocument.prototype.cast()` as an alternative for instantiating a subdocument without an associated parent document, which is the root cause of the issue in #14748.

**Examples**

### CODE REVIEW

File: test/types/schema.test.ts [hunk 1]
```diff
@@ -1555,3 +1556,17 @@ function gh14696() {
...
+  const schema = new Schema({
+    arr: [nestedSchema],
+    singleNested: nestedSchema
+  });
+
+  // Cast to single nested subdoc
+  const subdoc = schema.path('singleNested')
+    .cast<HydratedArraySubdocument<{ name: string }>>({ name: 'bar' });
+  expectAssignable<{ name: string }>(subdoc);
+}
```
Comment (hasezoey) [inline test/types/schema.test.ts:1572]:
```suggestion
}

```

fix lint

File: test/types/schema.test.ts [hunk 2]
```diff
@@ -1555,3 +1556,17 @@ function gh14696() {
...
+
+  const schema = new Schema({
+    arr: [nestedSchema],
+    singleNested: nestedSchema
+  });
+
+  // Cast to single nested subdoc
+  const subdoc = schema.path('singleNested')
+    .cast<HydratedArraySubdocument<{ name: string }>>({ name: 'bar' });
+  expectAssignable<{ name: string }>(subdoc);
```
Comment (hasezoey) [inline test/types/schema.test.ts:1571]: slight suggestion, also add a test line for the non-set return type overload

**Review by hasezoey — APPROVED** [review]
Looks good, some slight suggestions

**Comment by vkarpov15** [comment]
@hasezoey please take a look at the new implementation, I added a test based on your comments, and also some more logic to allow passing `DocType` down to the Schema Subdocument class

**Review by hasezoey — APPROVED** [review]

**Pull request merged**